### ISSUE TITLE
META: vendored code tracker

### ISSUE THREAD

**Issue by lucascolley** [scipy/scipy#21232]
I think that this will be a useful thing to have (to complement https://docs.scipy.org/doc/scipy/dev/core-dev/index.html#vendored-code) for a few reasons below. I thought that I'd open this before writing out the tracker in case we would prefer a different format (wiki page, docs only etc.).

1. There is less maintenance overhead for an issue than a docs page.
2. We don't have an exhaustive list yet - I only accidentally realised that we had a vendored copy of `numpydoc.docscrape` the other day.
3. It would be useful to categorise which code we do/don't maintain a fork of and maintain a list of upstream repos.
4. This will be a useful place to track potential progress from upstreams towards support for meson subprojects.

**Comment by nickodell** [comment]
Hello, I performed a search of SciPy for README files, using the command `find . -name README\*` and filtered out the ones which did not seem to correspond to external software.

List of additional vendored software:

- scipy/optimize/tnc
- scipy/optimize/lbfgsb_src
- scipy/optimize/_direct
- scipy/io/_fast_matrix_market
- scipy/interpolate/fitpack
- scipy/stats/libnpyrandom
- scipy/_lib/highs

Of note:

* It seems like `scipy/stats/libnpyrandom` is part of NumPy.
* Someone seems to be maintaining a modern Fortran version of `scipy/interpolate/fitpack` at https://github.com/perazz/fitpack.
* Highs is being maintained at https://github.com/ERGO-Code/HiGHS and seems pretty active.

**Comment by dschmitz89** [comment]
- DIRECT is taken from [nlopt](https://github.com/stevengj/nlopt) and maintained in the sense that bugs will likely be fixed but no new features implemented as NLopt also has its own DIRECT implementation.
- L-BFGS-B is taken from [here](https://users.iems.northwestern.edu/~nocedal/lbfgsb.html) and de facto unmaintained

**Comment by lucascolley** [comment]
Thanks @nickodell ! Added above.

And thanks @dschmitz89 ! Please feel free to update the tracker and sort any of the entries from the "to be audited" section.

A few findings:

- The URL in the TNC README (http://www.jeannot.org/~js/code/index.en.html) is dead
- `scipy.interpolate` README has
```
Author: Enthought, Inc
	Austin, TX
	enthough-dev@mail.enthought.com
```
perhaps time to change to
```
Authors:
    SciPy developers
        https://discuss.scientific-python.org/c/contributor/scipy/
    Enthought, Inc
        Austin, TX
        enthough-dev@mail.enthought.com
```
?

**Comment by ev-br** [comment]
scipy.interpolate README file can be just removed. 
Fitpack is not really vendored, what we carry has been heavily modified w.r.t. the netlib version over time. 

And the modernized implementation, who knows if it really works and what bugs it has ---  heisencode, as it were :-).

**Comment by lucascolley** [comment]
@tirthasheshpatel is there any vendored code under `scipy/stats/_unuran`, or is that all written by you, with the vendored code just in the git submodule?

**Comment by tirthasheshpatel** [comment]
> is there any vendored code under `scipy/stats/_unuran`, or is that all written by you, with the vendored code just in the git submodule?

Yes, none of the code under `scipy/stats/_unuran` is vendored, all the vendored code present in the submodule.

**Comment by h-vetinari** [comment]
> Git submodules, directly from active upstream
> * `scipy/sparse/linalg/_propack/PROPACK`: https://github.com/scipy/PROPACK

I don't think this is categorized correctly -- this is vendored code AFAIU. Not sure if there's a better upstream to point to than http://sun.stanford.edu/~rmunk/PROPACK/, but it's hardly "active".

**Comment by lucascolley** [comment]
Feel free to split that and scipy-mathjax out into a separate category

**Comment by lucascolley** [comment]
> it's hardly "active"

There are two different relevant senses of active.

1. Is there an out-of-SciPy-tree place where bug/security patches can be contributed?
2. Is there an upstream that is receiving somewhat regular updates and releases?

(1) is useful for categorisation here, but you are correct that (2) is much more typical use of "active" (and also useful for categorisation).

**Comment by dschmitz89** [comment]
A good candidate for vendored code refactoring would be `SuperLU` imo. It is widely available in distros and our patches are worth upstreaming in my opinion (mostly avoiding crashes with singular matrices).
